src/main/java/frc/robot/utils/MovingAverage.java

```diff
@@ -0,0 +1,114 @@
+
+import java.io.*;
+import java.util.Map;
+import java.util.TreeMap;
+
+/**
+ * The moving average contains a list of velocity based on the distance, and should give an approximation of the velocity to apply.
+ * In order to the moving average object to approximate better, you need to pass more values.
+ * The object calculate calculate a straight line between the closest point, and retrieve the velocity (in the straight line between the records).
+ */
+public class MovingAverage {
+    private final String DELIMITER = ",";
+    private final Map<Double, Double> distanceVelocityMap; // <key - distance, value - velocity>, tree map is automatically sort the values by distance.
+
+    public MovingAverage(String pathToCsv) {
+        this(new File(pathToCsv));
+    }
+
+    public MovingAverage(File csv) {
+        this.distanceVelocityMap = new TreeMap<>();
+        if (csv.exists() && csv.isFile()) {
+            try {
+                read(new FileReader(csv));
+            } catch (FileNotFoundException e) { // Can't happen, we have already checked it
+                e.printStackTrace();
+            }
+        } else
+            System.err.println("File not found: " + csv.getName());
+    }
+
+    public MovingAverage(Map<Double, Double> map) {
+        this.distanceVelocityMap = new TreeMap<>(map); // just because I want to order the values by the key.
+    }
+
+    public MovingAverage(double[][] values) {
+        this.distanceVelocityMap = new TreeMap<>();
+        for (double[] pair : values) {
+            if (pair.length >= 2)
+                this.distanceVelocityMap.put(pair[0], pair[1]);
+        }
+    }
+
+    public MovingAverage(Reader reader) {
+        this.distanceVelocityMap = new TreeMap<>();
+        read(reader);
+    }
+
+    private void read(Reader reader) {
+        String line = null;
+        try (BufferedReader bufferedReader = new BufferedReader(reader)) {
+            while ((line = bufferedReader.readLine()) != null) {
+                String[] pair = line.split(DELIMITER);
+                this.distanceVelocityMap.put(Double.parseDouble(pair[0]), Double.parseDouble(pair[1]));
+            }
+        } catch (IOException e) {
+            System.err.println("MovingAverage class can't read the file");
+            System.out.println(e.getMessage());
+        } catch (IndexOutOfBoundsException e) {
+            System.err.println("Please check that every line has a pair of values, the problem occurred on this line " + line);
+        }
+    }
+
+    /**
+     * Retrieve an approximation for the velocity based on the distance.
+     *
+     * @param distance the distance from the target.
+     * @return an approximation of the velocity to apply
+     */
+    public double estimateVelocity(double distance) {
+        double[][] nearest = getClosestDistance(distance);
+
+        if (nearest[0][0] == nearest[1][0]) // we have already have a record of the distance, just return the velocity based of that
```

Comment:
```suggestion
        if (nearest[0][0] == nearest[1][0]) // There is already a record of the distance, just return the velocity based of that
```
But I don't think you should do this check. In most cases, you won't use values that are in your table (vision processing or Kalman will probably return numbers with many digits), so this check will (almost) always return `true`.

And even if it didn't, you check that case twice: first in line 96 (`if (difference == 0)`) and then on this line.
I have an idea where you won't need any of these checks. If you don't find your own solution I'll share mine with you.

Reply:
I don't understand what do you mean with this statement

Reply:
Which part(s)? The first, second or both?

Reply:
the first

Reply:
You put this condition so if this method (`estimateVelocity`) receives a distance that is already in the table (=map), you'll return the corresponding velocity. Right?
What I'm saying is that it'll (almost) never happen, because the distances will come from vision processing or Kalman and these distances won't be in your table. Say you put in the table distances with jumps of 0.5 (`0.5, 1, 1.5, 2...`), on the field you won't have "pretty numbers" like these. You'd have long numbers with many digits.

Reply:
How?

Reply:
Think! 🙂
If you won't find a solution by tomorrow, I'll share mine with you.

Reply:
🕐🕑🕒🕓🕔🕕🕖🕗🕘🕙🕚🕛

Reply:
🙂
It seemed to me you've already fixed it in a66d93e732b2802de5bb4ef3b70ac3b8f4f515e1, but now I'm actually not so sure what were the changes there...
Anyway, the code is different now but my solution back then was that instead of that:
https://github.com/Galaxia5987/robot-2021/blob/c7e35481d185060d556b6bf6500e38a7d36ca20a/src/main/java/frc/robot/utils/MovingAverage.java#L96-L102
You'll do that:
```java
    if (difference == 0) {
            closest[0][0] = currentDistance;
            closest[0][1] = velocity;
            closest[1][0] = currentDistance + 0.1;
            closest[1][1] = velocity;
            return;
        }
```
Then, here:
https://github.com/Galaxia5987/robot-2021/blob/310444ce8aab8564e0729141972a096632071729/src/main/java/frc/robot/utils/MovingAverage.java#L70-L74
`higherVelocity == lowerVelocity` so `higherVelocity - lowerVelocity == 0` and `distances[1] - distances[0] == 0.1`. That means `m = 0`, and then:
https://github.com/Galaxia5987/robot-2021/blob/310444ce8aab8564e0729141972a096632071729/src/main/java/frc/robot/utils/MovingAverage.java#L74-L76
The returned value will be simply `b`, and `b == higherVelocity` if `m == 0`. This is the wanted behaviour, without a condition in this function.
(P.S: `closest[1][0] = currentDistance + 0.1` is needed to avoid division by zero, you can use here any value, `0.1` is an arbitrary choice.)
The code now is a bit different, but I think you can still implement my idea.
And even if you can't now, I enjoyed writing it and hope you learned something new 🙂.